Use UIDs in addition to paths for extracted meshes, materials and animations

lyuma opened the pull request:
Fixes #68672

edit 2025-02-12: rewrote this PR from scratch.
edit 2025-02-16: migrate existing UIDs, use hash function to derive consistent UIDs for meshes and animations.

This doesn't necessarily fix a bug, but this is trying to migrate more things to support UIDs. In this case, the extracted file paths for meshes and animations did not support UID, so if the paths were renamed it was possible for the import to fail.

Similarly, renaming materials could cause them to get lost between imports. This change will keep the uid updated so that the material can be referenced even after it is renamed.

This PR adds `use_external/fallback_path` and `save_to_file/fallback_path` to the existing `use_external/path` and `save_to_file/path` properties. These properties already use the ResourceUID::ensure_path() function which means they support UIDs already. The main change is to handle the case where the UID lookup fails (is_empty()) and to regenerate the UID from the fallback path.
Additionally, refresh `fallback_path` properties on every import to ensure that the path is kept up to date.
Finally, this changes Material, Mesh and Animation extraction buttons in the SceneImportSettings dialog to generate UIDs and assign the fallback_path initially.

Also, fix a bug in CSV translations related to UID hashing (as this was the only other place in the code which used a hash function to construct a UID, but did so incorrectly (allowing negative numbers) and without checking for collision.

fire reviewed (commented):
Are you sure we are ok to merge this if the other parts are not implemented? There's also a call for all uid code contributors to meetup.

lyuma commented:
@fire if you think the change is good and just needs the missing pieces implemented, such as animation slices, I will implement the missing parts next week. I just wanted get approval for the overall concept before feature freeze.

fire commented:
It makes sense that uid works everywhere.

makidoll commented:
Thanks for the PR, I was having this exact issue. However, I noticed if I select an external material under mesh import in the UI, the res:// path is shown but the uid:// path is saved in the .import file. You mentioned that "the UID path isn't obvious and readable" and got confused when the engine deals with this issue

Jordyfel commented:
Fixes #95637

The path import option can currently be set not only in `_save_dir_confirm()`, which is called only when the batch extraction dialog is used, but also by the custom inspector thing which is the `SceneImportSettingsData` class in the same file.

What I'm thinking is that a user wouldn't really need to change the path after it was set, they would just move the extracted file in the filesystem instead.

This could justify something like, instead of introducing a new uid import option, store the uid in the currently existing path import option and use `ResourceUID.ensure_path()` for compatibility. If the user wants to reset the path and extract again, they would set the `enabled` option to false and use the other dialog (through `_save_dir_confirm()`) again. Following this logic it also makes sense to hide the import option so that it is only saved but not editable in the custom inspector thing.

akien-mga commented:
Would be good to rebase and test again as #101256 was fixed.

KoBeWi commented:
The path and uid are now stored correctly.
However if you reimported files in 4.4 and they have uid in path field, it can't be fixed by simply reimporting; you need to extract to another file.

KoBeWi commented on editor/import/3d/scene_import_settings.cpp:
```diff
@@ -1590,6 +1590,7 @@ void SceneImportSettingsDialog::_save_dir_confirm() {
 
 				md.settings["use_external/enabled"] = true;
 				md.settings["use_external/path"] = path;
+				md.settings["use_external/uid"] = ResourceUID::get_singleton()->id_to_text(ResourceLoader::get_resource_uid(path));
```
```suggestion
				md.settings["use_external/uid"] = ResourceUID::path_to_uid(path);
```

KoBeWi commented on editor/import/3d/resource_importer_scene.cpp:
```diff
@@ -1528,8 +1529,26 @@ Node *ResourceImporterScene::_post_fix_node(Node *p_node, Node *p_root, HashMap<
 								String path = matdata["use_external/path"];
+								ResourceUID::ID uid = ResourceUID::INVALID_ID;
+								if (matdata.has("use_external/uid")) {
+									String uid_str = matdata["use_external/uid"];
+									uid = resource_uid->text_to_id(uid_str);
```
```suggestion
									uid = resource_uid->text_to_id(matdata["use_external/uid"]);
```

lyuma commented:
So from looking into this, my understanding is @reduz PRs #97912 and #97363 address a lot of the issues that this PR was designed for. Here are some of the things I am still thinking about:

one of the things I am considering changing is to keep status quo with path = "uid://" (or res:// for compatibility) and put the file path into a property called "fallback_path" only to be used when the uid does not exist. That would make this PR smaller and less dangerous compared to beta, since it is only fixing the case where you have extracted files and the referenced file was deleted.

the other question I had is when is it appropriate to modify the .import settings, for example if a referenced file is moved. If we call it fallback_path... let's say the user extracts a material to "res://models/materials/red.material" and then moves it to "res://global/red.material". Should the fallback_path be updated to the new path next time it is imported. Then if the user deletes global/red.material and the gltf is reimported, what should happen?

Finally, we need to make sure all places that extract resources are updated to use uid:// at the time they perform the extraction (for example in the adv import dialog), and I should test if animation slices work)

lyuma committed:
Scene import: extract UID paths and store fallback

When extracting meshes, materials and animations, always store the uid:// path as well as a res:// fallback.
When validating import settings, load the fallback path if the uid:// path fails to load.
Update save_to_file/fallback_path every import to keep the file path in sync with the uid.
Use UID hashing for meshes and animations.

lyuma commented:
Pushed a new version of this change. I updated the original comment to explain what the PR is now doing, but I wanted to have this ready for the core meeting. My goal is to have a minimal change. This does not change anything to users who use res:// paths in .import, and it does not change anything for users who use valid uid:// paths in .import.

During import, this PR should only impact cases when a uid:// path in .import is missing.

This PR also implements UID paths when the user selects a new directory to extract meshes, textures and materials, so in other words, it means that this code may also affect newly imported assets.

KoBeWi commented on editor/import/3d/scene_import_settings.cpp:
```diff
@@ -1585,9 +1585,18 @@ void SceneImportSettingsDialog::_save_dir_confirm() {
 		}
+		if (path.begins_with("uid://")) {
+			uid_path = path;
+			path = ResourceUID::ensure_path(uid_path);
```
```suggestion
			path = ResourceUID::uid_to_path(uid_path);
```
`ensure_path()` should be used when the original path can be either.

KoBeWi commented:
I duplicated a GLB file and extracted mesh, but it saved as regular path for `path` and empty `fallback_path`. Seems like it's unreliable?

I think the general idea is fine.

lyuma commented:
@KoBeWi I got this change working for meshes, materials and animations. I have been doing some tests on the fox gltf (which has animations, a mesh and a material) and so far I'm pretty satisfied with it.

It feels nice and transparent to the user, since the file pickers only show the res:// path. I tested renames, even in cases where Godot doesn't update the res:// path, and it seems to be able to locate the file fine using the uid://

Some known issues:
1. I do not currently upgrade res:// paths to uid:// - This is kind of intentional to reduce the scope of this PR and reduce risk, but I'm open to changing it, for example due to the second issue.
2. Due to how `PROPERTY_HINT_SAVE_FILE` seems to work, if a new file is created using the file picker in the inspector, it puts a `res://` path into `save_to_file/path` which won't get upgraded. (The common flow of using extract / set save paths buttons works correctly)
3. The `ResourceSaver::set_uid` API is jank both in potentially causing glitches and performance, and from my research, there is no other way to save a resource to a file with a pre-determined uid:

`ResourceSaver::save` doesn't have a way to pass a UID, and I saw the code @reduz linked to first calls `ResourceSaver::save`, and then uses `ResourceSaver::set_uid`. However, set_uid is a really bad function for a few reasons:

a. it rewrites the file again, so it now means all meshes and animations are being written twice.

b. the temporary uid is saved to uid_cache.bin first, which causes other places that reference the extracted meshes to use the temporary uid, even though the .tres file itself has a different UID. It's very confusing and while I haven't run into a bug with this, it feels wrong and jank. I think the fix requires modifying ResourceSaver to allow passing a UID into save(), which will turn into a bigger change.

lyuma commented on editor/import/3d/resource_importer_scene.cpp:
```diff
@@ -1804,16 +1818,21 @@ Ref<Animation> ResourceImporterScene::_save_animation_to_file(Ref<Animation> ani
+	ERR_FAIL_COND_V_MSG(err != OK, anim, "Saving of animation failed: " + res_path);
+	if (p_save_to_path.begins_with("uid://")) {
+		// slow
+		ResourceSaver::set_uid(res_path, ResourceUID::get_singleton()->text_to_id(p_save_to_path));
```
I have made the change for Mesh and Animation to use hash for UID generation.
(Note: The hash generation in CSV Translation had a bug that was causing uid conflicts and uid://<invalid>, which I opted to fix in this PR. Let me know if the translation fixes should be a separate PR)

For Material, it is common for people to use other Material objects, switch to ShaderMaterial, or share materials between models. I think randomly generated UID makes more sense for this workflow because they are not overwritten on each import.

KoBeWi commented on editor/import/3d/resource_importer_scene.cpp:
```diff
@@ -2044,6 +2068,7 @@ void ResourceImporterScene::get_internal_import_options(InternalImportCategory p
 		case INTERNAL_IMPORT_CATEGORY_MESH: {
 			r_options->push_back(ImportOption(PropertyInfo(Variant::BOOL, "save_to_file/enabled", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_DEFAULT | PROPERTY_USAGE_UPDATE_ALL_IF_MODIFIED), false));
 			r_options->push_back(ImportOption(PropertyInfo(Variant::STRING, "save_to_file/path", PROPERTY_HINT_SAVE_FILE, "*.res,*.tres"), ""));
+			r_options->push_back(ImportOption(PropertyInfo(Variant::STRING, "save_to_file/fallback_path", PROPERTY_HINT_SAVE_FILE, "", PROPERTY_USAGE_NO_EDITOR), ""));
```
```suggestion
			r_options->push_back(ImportOption(PropertyInfo(Variant::STRING, "save_to_file/fallback_path", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_NO_EDITOR), ""));
```
Hints are useless if the property is not editable.

KoBeWi commented on editor/import/3d/resource_importer_scene.cpp:
```diff
@@ -2196,17 +2224,17 @@ bool ResourceImporterScene::get_internal_option_visibility(InternalImportCategor
 		} break;
 		case INTERNAL_IMPORT_CATEGORY_MESH: {
-			if (p_option == "save_to_file/path") {
+			if (p_option == "save_to_file/path" || p_option == "save_to_file/fallback_path") {
```
Fallback path is always invisible, so this is unnecessary.

KoBeWi commented on editor/import/3d/scene_import_settings.cpp:
```diff
@@ -1585,9 +1585,13 @@ void SceneImportSettingsDialog::_save_dir_confirm() {
 		}
 		String path = item->get_text(1);
+		String uid_path = path;
 		if (!path.is_resource_file()) {
```
`uid://` will return false.

lyuma replied:
Ah I had the if statements in the wrong order. I didn't notice because the first time resources are extracted, they are always res://
Also found a related bug with material uid assignment. I realized for materials we can simply use the standard flow of ResourceSaver::save() -> ResourceUID::path_to_uid(). No need to complicate things by generating a UID first and then calling set_uid.

The material logic in scene_import_settings should be much simpler now

KoBeWi commented on editor/import/3d/resource_importer_scene.cpp:
```diff
@@ -1534,8 +1534,26 @@ Node *ResourceImporterScene::_post_fix_node(Node *p_node, Node *p_root, HashMap<
 							if (matdata.has("use_external/enabled") && bool(matdata["use_external/enabled"]) && matdata.has("use_external/path")) {
 								String path = matdata["use_external/path"];
 								Ref<Material> external_mat = ResourceLoader::load(path);
+								if (!external_mat.is_valid()) {
```
```suggestion
								if (external_mat.is_null()) {
```

KoBeWi commented:
I tested that it works (as much as I'm familiar with model import at least).
@reduz will probably have more to say about implementation.

AThousandShips commented on editor/import/3d/resource_importer_scene.cpp:
```diff
@@ -2584,16 +2612,24 @@ Node *ResourceImporterScene::_generate_meshes(Node *p_node, const Dictionary &p_
-					ResourceSaver::save(mesh, save_to_file); //override
+					Error err = ResourceSaver::save(mesh, save_res_path); //override
+					if (err != OK) {
+						WARN_PRINT(vformat("Failed to save mesh %s to %s", mesh->get_name(), save_res_path));
```
```suggestion
						WARN_PRINT(vformat("Failed to save mesh %s to '%s'.", mesh->get_name(), save_res_path));
```

akien-mga commented:
Thanks!

OhiraKyou commented:
Related? https://github.com/godotengine/godot/issues/74612
